### PR TITLE
fix(marshal): reject getters in pass-by-ref, even if it returns a function

### PR DESCRIPTION
**Pull request by warner**
This doesn't pass the typescript check yet, because it complains that I'm using a (potential) Symbol to do a property lookup:

```js
  const descs = getOwnPropertyDescriptors(val);
  const keys = ownKeys(descs); // enumerable-and-not, string-or-Symbol
  keys.forEach(key => {
    assert(!descs[key].get, X`cannot serialize objects with getters like ${q(String(key))} in ${val}`);
    assert.typeof(
      val[key],
      'function',
      X`cannot serialize objects with non-methods like ${q(
        String(key),
      )} in ${val}`,
    );
  });
```

But that's exactly what I need to do: I want to reject getters in Symbol-named properties too. I don't need to look at the property name (pass-by-reference doesn't care), so want something that returns key+value tuples like `Object.entries().forEach([_key, desc] ..`, except that `Object.entries` ignores Symbol-named properties, as does `Object.values()`.

Any ideas how I can appease typescript?

### CODE REVIEW

**Comment by dckc** [comment]
I suggest using `entries` rather than `keys` and `descs[key]`:

```js
  const descs = getOwnPropertyDescriptors(val);
  Object.entries(descs).forEach(([key, desc]) => {
    assert(!desc.get, X`cannot serialize objects with getters like ${q(String(key))} in ${val}`);
```

File: packages/marshal/src/marshal.js [hunk 1]
```diff
@@ -345,8 +345,14 @@ function assertCanBeRemotable(val) {
   assert(!Array.isArray(val), X`Arrays cannot be pass-by-remote`);
   assert(val !== null, X`null cannot be pass-by-remote`);
 
-  const keys = ownKeys(val);
+  const descs = getOwnPropertyDescriptors(val);
+  const keys = ownKeys(descs); // enumerable-and-not, string-or-Symbol
   keys.forEach(key => {
+    // @ts-ignore
+    assert(
+      !descs[key].get,
```
Comment (erights) [inline packages/marshal/src/marshal.js]: Not a good test. An accessor property with only a setter and no getter with have a descriptor with `get: undefined` which is falsy and so will illegitimately pass this test.
```suggestion
      !('get' in descs[key]),
```

Comment (erights) [inline packages/marshal/src/marshal.js]:
```suggestion
      !('get' in descs[/** type{string}*/ (key)]),
```
Somewhere recently I just came across this cast in some of our code. I did the @ts-ignore trick myself, only just recently but before I saw this. If this cast is the best way to do it, I should track down my `@ts-ignore`s for this and fix them,

It makes NO SENSE for typescript to prohibit indexing by symbols. Using symbols as property names is the use case for which symbols were introduced! And you can only use them as property names by indexing syntax. Attn @michaelfig is there something about this prohibition that I'm not getting?

File: packages/marshal/test/test-marshal.js [hunk 2]
```diff
@@ -287,21 +288,25 @@ test('records', t => {
       if (opt === 'enumStringData') {
         props.key1 = { enumerable: true, value: 'data' };
       } else if (opt === 'enumStringFunc') {
-        props.key2 = { enumerable: true, value: () => 0 };
-      } else if (opt === 'enumStringGet') {
-        props.key3 = { enumerable: true, get: () => 0 };
+        props.enumStringFunc = { enumerable: true, value: () => 0 };
```
Comment (erights) [inline packages/marshal/test/test-marshal.js:291]: If you're going to give the string-named properties meaningful string names, which is nice, you could also give the symbols above similar meaningful names.

**Review by erights — APPROVED** [review]
LGTM

**Comment by erights** [comment]
> I suggest using `entries` rather than `keys` and `descs[key]`:
> ...

@dckc Using `entries` was exactly the mistake I made that set us down this long road. What I did not understand is how restrictive `entries` is. It only enumerates *string-named* enumerable own properties. It does *not* enumerate symbol-named own properties.

File: packages/marshal/src/marshal.js [hunk 3]
```diff
@@ -345,8 +345,14 @@ function assertCanBeRemotable(val) {
   assert(!Array.isArray(val), X`Arrays cannot be pass-by-remote`);
   assert(val !== null, X`null cannot be pass-by-remote`);
 
-  const keys = ownKeys(val);
+  const descs = getOwnPropertyDescriptors(val);
+  const keys = ownKeys(descs); // enumerable-and-not, string-or-Symbol
   keys.forEach(key => {
+    // @ts-ignore
```
Comment (erights) [inline packages/marshal/src/marshal.js]: I see from the CI output that this `@ts-ignore` isn't working for you anyway. I suspect it might if you move it down a line, but see below instead.

**Review by erights — APPROVED** [review]

**Comment by erights** [comment]
> > I suggest using `entries` rather than `keys` and `descs[key]`:
> > ...
> 
> @dckc Using `entries` was exactly the mistake I made that set us down this long road. What I did not understand is how restrictive `entries` is. It only enumerates _string-names_ enumerable own properties. It does _not_ enumerate symbol-named own properties.

The new comments at 
https://github.com/Agoric/agoric-sdk/blob/97c1bc39e6de1986dcdc20f64d5796d8ecbad887/packages/zoe/src/objArrayConversion.js#L59-L79
is a good example both of how tricky all this is in JavaScript, and how great it is that our pass-by-copy objects avoid these hazards.

File: packages/marshal/src/marshal.js (hunk 1 above)
Comment (warner) [reply packages/marshal/src/marshal.js]: added

Comment (warner) [reply packages/marshal/src/marshal.js]: I tried that inline `type` and it didn't seem to work. I broke down and moved the `// @ts-ignore` into the middle of the multi-line `assert()` clause and that did the trick.

File: packages/marshal/test/test-marshal.js (hunk 2 above)
Comment (warner) [reply packages/marshal/test/test-marshal.js:291]: done

**Review by erights — APPROVED** [review]
The changes look good

File: packages/marshal/src/marshal.js (hunk 1 above)
Comment (michaelfig) [reply packages/marshal/src/marshal.js]: > It makes NO SENSE for typescript to prohibit indexing by symbols. Using symbols as property names is the use case for which symbols were introduced! And you can only use them as property names by indexing syntax. Attn @michaelfig is there something about this prohibition that I'm not getting?


TL;DR: This is a bug in Typescript that was noticed in 2015 but has never been corrected.  It has a "help wanted" tag on it.

You can have some fun reading at https://github.com/microsoft/TypeScript/issues/1863 (100+ comments).
